cognitive/src/main/scala/com/microsoft/ml/spark/cognitive/TextTranslator.scala

```diff
@@ -0,0 +1,406 @@
+
+  override protected def contentType: Row => String = { _ => "application/json; charset=UTF-8" }
+
+  override protected def inputFunc(schema: StructType): Row => Option[HttpRequestBase] = {
+    val rowToUrl = prepareUrl
+    val rowToEntity = prepareEntity;
+    { row: Row =>
+      if (shouldSkip(row)) {
+        None
+      } else {
+        val req = prepareMethod()
+        req.setURI(new URI(rowToUrl(row)))
+        getValueOpt(row, subscriptionKey).foreach(
+          req.setHeader(subscriptionKeyHeaderName, _))
+        getValueOpt(row, subscriptionRegion).foreach(
+          req.setHeader(subscriptionRegionHeaderName, _)
+        )
+        req.setHeader("Content-Type", contentType(row))
+
+        req match {
+          case er: HttpEntityEnclosingRequestBase =>
+            rowToEntity(row).foreach(er.setEntity)
+          case _ =>
+        }
+        Some(req)
+      }
+    }
+  }
+
+  override protected def prepareUrl: Row => String = {
+    val urlParams: Array[ServiceParam[Any]] =
+      getUrlParams.asInstanceOf[Array[ServiceParam[Any]]];
+
+    // This semicolon is needed to avoid argument confusion
+    def replaceName(s: String): String = {
+      if (s == "fromLanguage") {
+        "from"
+      } else if (s == "toLanguage") {
+        "to"
+      } else {
+        s
+      }
+    }
+    { row: Row =>
+      val base = getUrl + "?api-version=3.0"
+      val appended = if (!urlParams.isEmpty) {
+        "&" + URLEncodingUtils.format(urlParams.flatMap(p =>
+          getValueOpt(row, p).map {
+            v => replaceName(p.name) -> p.toValueString(v)
+          }
+        ).toMap)
+      } else {
+        ""
+      }
+      base + appended
+    }
+  }
+
+  override def setLocation(v: String): this.type = {
+    setSubscriptionRegion(v)
+    setUrl("https://api.cognitive.microsofttranslator.com/" + urlPath)
+  }
+
+}
+
+object Translate extends ComplexParamsReadable[Translate]
+
+class Translate(override val uid: String) extends TextTranslatorBase(uid)
+  with TextAsOnlyEntity with BasicLogging {
+  logClass()
+
+  def this() = this(Identifiable.randomUID("Translate"))
+
+  def urlPath: String = "translate"
+
+  val toLanguage = new ServiceParam[Seq[String]](this, "toLanguage", "Specifies the language of the output" +
+    " text. The target language must be one of the supported languages included in the translation scope." +
+    " For example, use to=de to translate to German. It's possible to translate to multiple languages simultaneously" +
+    " by repeating the parameter in the query string. For example, use to=de&to=it to translate to German and Italian.",
+    isRequired = true, isURLParam = true,
+    toValueString = { seq => seq.mkString(",") })
+
+  def setToLanguage(v: Seq[String]): this.type = setScalarParam(toLanguage, v)
+
+  def setToLanguageCol(v: String): this.type = setVectorParam(toLanguage, v)
+
+  val fromLanguage = new ServiceParam[String](this, "fromLanguage", "Specifies the language of the input" +
```

Comment:
Can abstract the fromLanguage and toLanguage methods into separate traits and "mix together" to reduce code. Also be on the lookout for other params that can be "factored" out in this manner as its less maintenance work for us later on ;)

Reply:
Although there're several fromLanguage & toLanguage in this file, the 'isRequired' parameter and type might be different, so didn't factor these two within Translate out. Or do you have other approaches to mix them together?

Reply:
Okay! One idea would be to add protected fromLanguageRequired: Boolean = true on base class and just override that